lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -568,176 +569,188 @@ def write(chunk)
           tag, group_level_resource)}"
 
         # Does the actual write to the cloud logging api.
-        client = api_client
         if @use_grpc
-          begin
-            labels_utf8_pairs = group_level_common_labels.map do |k, v|
-              [k.encode('utf-8'), convert_to_utf8(v)]
-            end
+          labels_utf8_pairs = group_level_common_labels.map do |k, v|
```

Comment:
Since we're decomposing the request in `write_request_via_grpc` anyway, would it make sense to fold these tranformations into `write_request_via_grpc` and send in the same request object as on the HTTP path? We could even set `partial_success` on that one and ignore it (with a comment) in `write_request_via_grpc`.

Reply:
The utf8 transformation has been folded into `write_request_via_grpc`. Also made the changes to pipe up requests here in REST format and let gRPC path decompose them later.

Reply:
In fact, this is a bit weird now... We are send the same request (in REST format) for both REST and gRPC. But within the request, the entries still have different format.

Reply:
I just refactored that part to use callers to construct both entries and requests.

Reply:
Yeah, I agree it's not the most intuitive approach. Now that we only have the single `request` argument, would it make sense to have an array of hashes, send it in as a bunch of decomposed pieces and reconstruct in both methods? E.g.:
```ruby
        requests_to_send << {
          entries: entries,
          log_name: log_name,
          resource: group_level_resource,
          labels: group_level_common_labels,
          partial_success: @partial_success
        }
      end
      # Does the actual write to the Stackdriver Logging API.
      write_request = method(if @use_grpc
                               :write_request_via_grpc
                             else
                               :write_request_via_http
                             end)
      requests_to_send.each do |request|
        write_request.call(**request)
      end
...


    def write_request_via_grpc(entries:, log_name:, resource:, labels:,
                               partial_success:)
      client = api_client
      entries_count = entries.length
      utf8_encoded_labels = labels.map do |k, v|
        [k.encode('utf-8'), convert_to_utf8(v)]
      end.to_h
      client.write_log_entries(
        # Ignore partial_success for gRPC path.
        entries,
        log_name: log_name,
        resource: Google::Api::MonitoredResource.new(
          type: resource.type,
          labels: resource.labels.to_h
        ),
        labels: utf8_encoded_labels
      )
      ...

    def write_request_via_http(entries:, log_name:, resource:, labels:,
                               partial_success:)
      client = api_client
      entries_count = entries.length
      client.write_entry_log_entries(
        Google::Apis::LoggingV2::WriteLogEntriesRequest.new(**request),
        options: { api_format_version: '2' }
      )
```

WDYT?

Reply:
That's an interesting idea. I gave it a try. Unfortunately `Keyword Arguments` is not supported by Ruby 2.1... Right now we support Ruby 2.1 - 2.4 (https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/blob/master/.travis.yml).

BTW, I made some change in commit f7337c56cd44546f39ed1cd7e658af33096d0cec to use callers.

Reply:
That's fine, you can pass in the hash and access its elements, to make it Ruby 2.1-compatible. Keyword arguments make the code nicer, but aren't essential for this solution.

Reply:
Cool cool. Just changed the code to pass a hash instead of a request object.